gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,482 @@
+        try:
+            self.ctx = mp.get_context(context)
+        except AttributeError:
+            logger.warn('Context switching for `multiprocessing` is not '
+                'available in Python 2. Using the default context.')
+            self.ctx = mp
+        self.env_fns = env_fns
+        self.shared_memory = shared_memory
+        self.copy = copy
+        self.max_retries = max_retries
+        self.episodic = episodic
+        self._num_retries = 0
+
+        if (observation_space is None) or (action_space is None):
+            dummy_env = env_fns[0]()
+            observation_space = observation_space or dummy_env.observation_space
+            action_space = action_space or dummy_env.action_space
+            dummy_env.close()
+            del dummy_env
+        super(AsyncVectorEnv, self).__init__(num_envs=len(env_fns),
+            observation_space=observation_space, action_space=action_space)
+
+        if self.shared_memory:
+            self._obs_buffer = create_shared_memory(
+                self.single_observation_space, n=self.num_envs)
+            self.observations = read_from_shared_memory(self._obs_buffer,
+                self.single_observation_space, n=self.num_envs)
+        else:
+            self._obs_buffer = None
+            self.observations = create_empty_array(
+                self.single_observation_space, n=self.num_envs, fn=np.empty)
+
+        self.parent_pipes, self.processes = [], []
+        self.error_queue = self.ctx.Queue()
+        with clear_mpi_env_vars():
+            for index in range(self.num_envs):
+                process, parent_pipe = self._start_process(index)
+                self.parent_pipes.append(parent_pipe)
+                self.processes.append(process)
+
+        self.waiting_reset = False
+        self.waiting_step = False
+        self._check_observation_spaces()
+
+    def seed(self, seeds=None):
+        self._assert_is_running()
+        if seeds is None:
+            seeds = [None for _ in range(self.num_envs)]
+        if isinstance(seeds, int):
+            seeds = [seeds + i for i in range(self.num_envs)]
+        assert len(seeds) == self.num_envs
+
+        for pipe, seed in zip(self.parent_pipes, seeds):
+            pipe.send(('seed', seed))
+        for pipe in self.parent_pipes:
+            pipe.recv()
+
+    def reset_async(self):
+        self._assert_is_running()
+        if self.waiting_step:
+            logger.warn('Calling `reset_async` while waiting for a pending '
+                'call to `step` to complete. Cancelling the previous call to '
+                '`step`...')
+            try:
+                self.step_wait(timeout=0)
+            except mp.TimeoutError:
+                pass
+
+        if self.waiting_reset:
+            logger.error('Calling `reset_async` while waiting for a pending '
+                'call to `reset` to complete. Closing `{0}`...'.format(self))
+            self.close(terminate=True)
```

Comment:
Why is closed called here? Shouldn't that be the job of the caller?

Reply:
The reason why I was calling `close` here (and elsewhere) was to avoid any hanging process if the user wasn't careful in closing the environment himself. But leaving the responsibility up to the user is also fine for me.